packages/core/src/result-transformers.ts

```diff
@@ -0,0 +1,190 @@
+   * along with the {@link ResultSummary} and {@link Result#keys}.
+   *
+   * NOTE: The config object requires map or/and collect to be valid.
+   *
+   * @example
+   * // Mapping the records
+   * const { keys, records, summary } = await driver.executeQuery('MATCH (p:Person{ age: $age }) RETURN p.name as name', { age: 25 }, {
+   *   resultTransformer: neo4j.resultTransformers.mappedResultTransformer({
+   *     map(record) {
+   *        return record.get('name')
+   *     }
+   *   })
+   * })
+   *
+   * records.forEach(name => console.log(`${name} has 25`))
+   *
+   * @example
+   * // Mapping records and collect result
+   * const names = await driver.executeQuery('MATCH (p:Person{ age: $age }) RETURN p.name as name', { age: 25 }, {
+   *   resultTransformer: neo4j.resultTransformers.mappedResultTransformer({
+   *     map(record) {
+   *        return record.get('name')
+   *     },
+   *     collect(records, summary, keys) {
+   *        return records
+   *     }
+   *   })
+   * })
+   *
+   * names.forEach(name => console.log(`${name} has 25`))
+   *
+   * @example
+   * // The transformer can be defined one and used everywhere
+   * const getRecordsAsObjects = neo4j.resultTransformers.mappedResultTransformer({
+   *   map(record) {
+   *      return record.toObject()
+   *   },
+   *   collect(objects) {
+   *      return objects
+   *   }
+   * })
+   *
+   * // The usage in a driver.executeQuery
+   * const objects = await driver.executeQuery('MATCH (p:Person{ age: $age }) RETURN p.name as name', { age: 25 }, {
+   *   resultTransformer: getRecordsAsObjects
+   * })
+   * objects.forEach(object => console.log(`${object.name} has 25`))
+   *
+   *
+   * // The usage in session.executeRead
+   * const objects = await session.executeRead(tx => getRecordsAsObjects(tx.run('MATCH (p:Person{ age: $age }) RETURN p.name as name')))
+   * objects.forEach(object => console.log(`${object.name} has 25`))
+   *
+   * @experimental
+   * @param {object} config The result transformer configuration
+   * @param {function(record:Record):R} [config.map=function(record) {  return record }] Method called for mapping each record
+   * @param {function(records:R[], summary:ResultSummary, keys:string[]):T} [config.collect=function(records, summary, keys) { return { records, summary, keys }}] Method called for mapping
+   * the result data to the transformer output.
+   * @returns {ResultTransformer<T>} The result transformer
+   * @see {@link Driver#executeQuery}
+   * @see https://github.com/neo4j/neo4j-javascript-driver/discussions/1052
+   */
+  mappedResultTransformer <
+    R = Record, T = { records: R[], keys: string[], summary: ResultSummary }
+  >(config: { map?: (rec: Record) => R, collect?: (records: R[], summary: ResultSummary, keys: string[]) => T }): ResultTransformer<T> {
+    if (config == null || (config.collect == null && config.map == null)) {
```

Comment:
so you allow a `null` map (if collect is define)? what happens then?

Reply:
not sure to understand what `state.records.push(record as unknown as R)` does

Reply:
> so you allow a null map (if collect is define)? what happens then?

The collect will receive the list of records, summary and keys for creating the final output. 

> not sure to understand what state.records.push(record as unknown as R) does

At that point, the record is typed as Record, however the generics is defining `R` (which is probably Record, but it could be something else if you force it). In theory, we could just do `record as R`. However, it's a good practice enforced by the linter to cast to `unknown` before. 

`unknown` is like `any`, the difference is that it causes type error when an unknown type value is assigned to a type which is not `any | unknown`, but an unknown variable can receive value of any types just like `any`